### DISCUSSION TITLE
Color-Key Mask

### DISCUSSION THREAD

**Discussion by jasonkylejacobs** [pymupdf/PyMuPDF#1259] [Q&A]
I'm currently trying to apply a simple color-key mask to a foreground JBIG2 image, so that I can make any white pixel transparent, however I can't seem to find a way to do this using PyMuPDF. I can achieve the desired result by creating a separate image mask, however this is a little overkill for what I need.

As a proof of concept for myself, these are the steps I've taken:

1) Create the PDF using a background and foreground image
```
doc = fitz.open()
page = doc.newPage()
bg_stream = open("wtr_0009_bg.jpg", 'rb').read()
fg_stream = open("wtr_0009_fg.jb2", 'rb').read()
page.insertImage(page.rect, stream=bg_stream, mask=None)
page.insertImage(page.rect, stream=fg_stream, mask=None)
doc.save("mask_test.pdf", deflate=True)
```

2) Edit the PDF using a text editor, to manually insert the color-key mask:
```
<<
/Filter 				/JBIG2Decode
/Type 					/XObject
/Subtype 				/Image
/BitsPerComponent 		1
/Width 					3421
/Height 				4904
/ColorSpace 			/DeviceGray
/Mask  					[1 1]
/Length 				6339
>>
```

This successfully makes all the white pixels in the foreground image transparent, the only downside is that the background image ends up distorted, although I'm assuming this is due to me editing the pdf with a text editor.

Is it currently possible to create the color-key mask directly using PyMuPDF?

If not, is this something that could be considered for a future release?
It would be great to be able to do something like:
`page.insertImage(page.rect, stream=fg_stream, mask=[1, 1])`

Thanks :)

**Comment by JorjMcKie** [comment]
Not 100% sure if this meets your requirements:
You can quite arbitrarily edit any PDF object using PyMuPDF. For example
```python
doc.xref_set_key(xref, "Mask", "[1 1]")
```
Editing across PDF dictionary key hierarchies also works. The following edits a page object such that some image reference name points to a different xref:
```python
doc.xref_set_key(page.xref, "Resources/XObject/Im0", "4711 0 R")
```

**Comment by JorjMcKie** [comment]
If you insert an image, the resulting xref is returned. You can inspect its details via `doc.xref_get_keys(xref)`, resp. each single key's value via `doc.xref_get_key(xref, "ColorSpace")`, etc.

**Comment by JorjMcKie** [comment]
For example:
```python
>>> import fitz
>>> doc=fitz.open()
>>> page=doc.new_page()
>>> xref=page.insert_image((100,100,200,200),filename="nur-ruhig.jpg")
>>> print(doc.xref_object(xref))
<<
  /Filter /DCTDecode
  /Type /XObject
  /Subtype /Image
  /BitsPerComponent 8
  /Width 439
  /Height 501
  /ColorSpace /DeviceRGB
  /Length 50784
>>
>>> for key in doc.xref_get_keys(xref):
	print(key, doc.xref_get_key(xref, key))

	
Filter ('name', '/DCTDecode')
Type ('name', '/XObject')
Subtype ('name', '/Image')
BitsPerComponent ('int', '8')
Width ('int', '439')
Height ('int', '501')
ColorSpace ('name', '/DeviceRGB')
Length ('int', '50784')
>>> 
```
